gpkit/geometric_program.py

```diff
@@ -1,380 +1,99 @@
-                         ["\\end{array}"])
-
-    def test(self, solver=None, printing=False, skipfailures=False):
-        end = " with default settings."
-        try:
-            self.solve(solver, printing, skipfailures)
-            print(self.__class__.__name__, "solved successfully" + end)
-        except Exception as e:
-            print(self.__class__.__name__, "failed to solve" + end)
-            self.checkbounds()
-            raise(e)
-
-    def solve(self, *args, **kwargs):
-        return self._solve(*args, **kwargs)
-
-    def _solve(self, solver=None, printing=True, skipfailures=False,
-               allownonoptimal=False):
+    def __init__(self, cost, constraints, verbosity=1):
+        self.cost = cost
+        self.constraints = constraints
+        self.posynomials = [cost] + list(constraints)
+        self.cs = np.hstack((mag(p.cs) for p in self.posynomials))
+        if not all(self.cs > 0):
+            raise ValueError("GeometricPrograms cannot contain coefficients"
+                             " less than or equal to zero.")
+        self.exps = functools_reduce(add, (x.exps for x in self.posynomials))
+        self.varlocs, self.varkeys = locate_vars(self.exps)
+        # k [j]: number of monomials (columns of F) present in each constraint
+        self.k = [len(p.cs) for p in self.posynomials]
+        # p_idxs [i]: posynomial index of each monomial
+        p_idx = 0
+        p_idxs = []
+        for p_len in self.k:
+            p_idxs += [p_idx]*p_len
+            p_idx += 1
+        self.p_idxs = np.array(p_idxs)
+
+        self.A, missingbounds = genA(self.exps, self.varlocs)
+
+        if verbosity > 0:
+            for var, bound in sorted(missingbounds.items()):
+                print("%s has no %s bound" % (var, bound))
+
+    def solve(self, solver=None, verbosity=1, *args, **kwargs):
         """Solves a GeometricProgram and returns the solution.
 
         Arguments
         ---------
-        printing : bool (optional)
-            If True (default), then prints out solver used and time to solve.
+        solver : str or function (optional)
+            By default uses one of the solvers found during installation.
+            If set to "mosek", "mosek_cli", or "cvxopt", uses that solver.
+            If set to a function, passes that function cs, A, p_idxs, and k.
+        verbosity : int (optional)
+            If greater than 0, prints solver name and solve time.
+        *args, **kwargs :
+            Passed to solver constructor and solver function.
+
 
         Returns
         -------
-        solution : dict
-            A dictionary containing the optimal values for each free variable.
+        result : dict
+            A dictionary containing the translated solver result.
         """
         if solver is None:
-            solver = self.solver
+            from . import settings
+            solver = settings['installed_solvers'][0]
```

Comment:
`__init__.py` sets `settings = {"installed_solvers": [""]}` in case of an `IOError` reading settings. Would it be cleaner to set `"installed_solvers": []`, and then move the conditional here to `if settings['installed_solvers']`?

(I'm just a little vexed about the `[0]` access relying on that empty string having been written).

Very minor, not urgent.